include/espi.h

```diff
@@ -0,0 +1,446 @@
+
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public APIs for eSPI driver
+ */
+
+#ifndef ZEPHYR_INCLUDE_ESPI_H_
+#define ZEPHYR_INCLUDE_ESPI_H_
+
+#include <misc/__assert.h>
+#include <misc/slist.h>
+
+#include <zephyr/types.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief ESPI Driver APIs
+ * @defgroup espi_interface ESPI Driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+/*
+ * eSPI VW channel packets contain 4 VWire states.
+ */
+#define ESPI_VW_SRC_ID0         0
+#define ESPI_VW_SRC_ID1         1
+#define ESPI_VW_SRC_ID2         2
+#define ESPI_VW_SRC_ID3         3
+#define ESPI_VW_SRC_ID_MAX      4
+
+
+/** @cond INTERNAL_HIDDEN */
+struct espi_vwire {
+	u8_t	idx;
+	u8_t	bit;
+	u8_t	dir;
+};
+
+/**
+ * @endcond
+ */
+
+/**
+ * @brief ESPI I/O mode capabilities
+ */
+enum espi_io_mode {
+	DUAL_LINES = BIT(0),
+	QUAD_LINES = BIT(1),
+	SINGLE_LINE = BIT(2),
+};
+
+/**
+ * @brief ESPI frequency capabilities
+ */
+enum espi_freq {
+	MAX_FREQ_20MHz,
+	MAX_FREQ_25MHz,
+	MAX_FREQ_33MHz,
+	MAX_FREQ_50MHz,
+	MAX_FREQ_66MHz
+};
+
+/**
+ * @brief ESPI channels support
+ */
+
+enum espi_channel {
+	PERIPHERAL_CHANNEL	= BIT(0),
+	VWIRE_CHANNEL		= BIT(1),
+	OOB_CHANNEL			= BIT(2),
+	FLASH_CHANNEL		= BIT(3),
+};
+
+enum espi_peripheral {
+	KEYBOARD_8042		= BIT(0),
+	ACPI				= BIT(1),
+	LEGACY_PORT92		= BIT(2),
+	UART				= BIT(3),
+	BIOS_PORT80			= BIT(4),
+};
+
+enum espi_bus_event {
+	ESPI_RESET = 1,
+	CHANNEL_READY,
+	VW_RECEIVED,
+	OOB_RECEIVED,
+};
+
+enum espi_vw_signal {
```

Comment:
Does this need to be exposed publicly? I mean, besides the master and slave driver, the user won't have to manipulate these afaik.

Anyway, add ESPI_VWIRE_SIGNAL_ as prefix to all values.

Reply:
Most of eSPI traffic including eSPI vwire signals are can be handled within the eSPI drivers but certain events are required to be exposed to user apps similar to GPIO control or events that trigger eSPI-unrelated flows such as power management, thermal management, reset sequence

+-------+                  +-------------+   +------+     +---------+
|  App  |                  | eSPI driver |   |  HW  |     |eSPI Host|
+---+---+                  +-------+-----+   +---+--+     +----+----+
    |                              |             |             |
    |   espi_init_callback         |             |             |
    +----------------------------> |             |             |
    |   espi_add_callback          |             |
    +----------------------------->+             |
    |                              |             |  VW CH ready|  eSPI host
    |                              |    IRQ      +<------------+  enables VW
    |                              | <-----------+             |  channel
    |                              |             |             |
    |                              | Processed   |             |
    |                              |             |             |
    |                              |             |             |
    |                              |             |             |
    |                              |             |             |
    |                              |             |             |
    |                              |             |             |
    |                              |             | HOST_RST    |  eSPI host
    |                              |             <-------------+  send VWire
    |                              +<------------+             |
    +<-----------------------------+ callback    |             |
    | App reset host-related       |             |             |
    | data structures              |             |             |
    |                              |             |             |
    |                              |             |   C10       |
    |                              |             +<------------+
    |                              <-------------+             |
    <------------------------------+             |             |
    | App executes                 |             |             |
    + power policy                 +             +             +